### PR TITLE
feat(stream): new boxed group_top_n executor from proto

### PR DESCRIPTION
**Pull request by Shmiwy**
Signed-off-by: Shmiwy <wyf000219@126.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

+ impl `ToStreamProst` `fmt::Display` and `PlanTreeNodeUnary` for `StreamGroupTopN`
+ impl `ExecutorBuilder` for GroupTopNExecutorBuilder

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
#4246 #3489

### CODE REVIEW

File: proto/stream_plan.proto [hunk 1]
```diff
@@ -174,6 +174,15 @@ message TopNNode {
   uint32 table_id_h = 6;
 }
 
+message GroupTopNNode {
+  repeated plan_common.ColumnOrder column_orders = 1;
+  uint64 limit = 2;
+  uint64 offset = 3;
+  repeated uint32 distribution_key = 4;
+  uint32 table_id_l = 5;
```
Comment (skyzh) [inline proto/stream_plan.proto:182]: Just call it table_id

File: src/frontend/src/optimizer/plan_node/stream_group_topn.rs [hunk 2]
```diff
@@ -111,3 +112,68 @@ impl StreamGroupTopN {
...
+            .iter()
+            .map(FieldOrder::to_protobuf)
+            .collect();
+        let group_key = self.group_key.iter().map(|idx| *idx as u32).collect();
+
+        let group_topn_node = GroupTopNNode {
+            column_orders,
+            limit: self.limit as u64,
+            offset: self.offset as u64,
+            distribution_key: vec![], // TODO: seems unnecessary
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/stream_group_topn.rs:131]: Then we can remove this field from proto?

**Review by skyzh — COMMENTED** [review]
Seems that we have not assigned table id on meta node?

File: src/frontend/src/optimizer/plan_node/stream_group_topn.rs [hunk 3]
```diff
@@ -111,3 +113,63 @@ impl StreamGroupTopN {
...
+    fn to_stream_prost_body(&self) -> ProstStreamNode {
+        use risingwave_pb::stream_plan::*;
+        let group_key = self.group_key.iter().map(|idx| *idx as u32).collect();
+
+        let group_topn_node = GroupTopNNode {
+            limit: self.limit as u64,
+            offset: self.offset as u64,
+            group_key,
+            table: Some(self.infer_internal_table_catalog().to_prost(
+                SchemaId::placeholder() as u32,
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/stream_group_topn.rs:127]:
```
.to_prost(
                    SchemaId::placeholder() as u32,
                    DatabaseId::placeholder() as u32,
                )
```

It seems that we are using such way everywhere. Shall we add a new function like `to_state_table_prost()` that uses these two placeholder ids by default?

File: src/frontend/src/optimizer/plan_node/stream_group_topn.rs [hunk 4]
```diff
@@ -111,3 +113,63 @@ impl StreamGroupTopN {
...
+                OrderDisplay {
+                    order: self.order(),
+                    input_schema
+                }
+            ),
+        );
+        builder
+            .field("limit", &format_args!("{}", self.limit))
+            .field("offset", &format_args!("{}", self.offset))
+            .field("group key", &format_args!("{:?}", self.group_key))
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/stream_group_topn.rs:154]: group_key

File: src/frontend/src/optimizer/plan_node/stream_group_topn.rs [hunk 5]
```diff
@@ -111,3 +113,63 @@ impl StreamGroupTopN {
         internal_table_catalog_builder.build(dist_keys, self.base.append_only)
     }
 }
+
+impl ToStreamProst for StreamGroupTopN {
+    fn to_stream_prost_body(&self) -> ProstStreamNode {
+        use risingwave_pb::stream_plan::*;
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/stream_group_topn.rs:119]: shall we panic when self.limit == 0?

File: src/frontend/src/stream_fragmenter/mod.rs [hunk 6]
```diff
@@ -334,8 +334,15 @@ impl StreamFragmenter {
             }
 
             NodeBody::TopN(top_n_node) => {
-                top_n_node.table_id_l = state.gen_table_id();
-                top_n_node.table_id_h = state.gen_table_id();
+                if let Some(table) = &mut top_n_node.table {
+                    table.id = state.gen_table_id();
+                }
```
Comment (skyzh) [inline src/frontend/src/stream_fragmenter/mod.rs:339]: else panic

File: src/meta/src/stream/stream_graph.rs [hunk 7]
```diff
@@ -616,14 +616,43 @@ impl StreamGraphBuilder {
...
                         node.table_id_l += table_id_offset;
                         node.table_id_h += table_id_offset;
 
-                        // TODO add catalog::Table to TopNNode
+                        // TODO add catalog::Table to AppendOnlyTopN
                         check_and_fill_internal_table(node.table_id_l, None);
                         check_and_fill_internal_table(node.table_id_h, None);
                     }
+                    NodeBody::TopN(node) => {
+                        if let Some(table) = &mut node.table {
```
Comment (skyzh) [inline src/meta/src/stream/stream_graph.rs:628]: Need refactor later. This pattern also appears everywhere.

```
                            table.id += table_id_offset;
                            table.schema_id = ctx.schema_id;
                            table.database_id = ctx.database_id;
```

**Review by skyzh — COMMENTED** [review]
Looks good for now. Will approve when all tests pass.

**Comment by Shmiwy** [comment]
@skyzh When all the tests pass, I will optimize the code and reduce code duplication

**Review by skyzh — APPROVED** [review]
LGTM for now. Need a new PR to implement rule to generate two-phase agg to ensure this change works.

**Pull request merged**